TykTechnologies/tyk
# Distributed Quota branch, cherry-picked against master

Ok this is a complicated one.

Distributed Quota feature (lets call it a beta): instead of having the quota enforced by a shared counter in redis, the counter is in-memory and counters are shared across a message bus to other cluster gateways, then merged, to create a distributed total. This means no centralised state and no one-to-one linkage with Redis.

**What we are trying to achieve:**

1. Stop having a one-to-one relationship with redis, when a counter-based quota is enforced, every inbound request means a roundtrip to redis, this is expensive and adds latency, even if done async it floods the redis connection counter pool and means performance degrades as traffic increases
2. Make counter operations faster: Since these are in-memory and are merged in the background, the counter is entirely in-memory and no round-trip or lookup is required. This should improve throughput for metered connections
3. The new DQ Quota object means that actually we can introduce new metering metrics because it is an abstraction
4. Managed redis writes (or any persistence), since the distributed cluster, with a single leader, will only write to the store every `x` interval (default to 3 seconds),  so no matter how high the load, the writes speed isn't a linear relation to throughput.

To enable set the following in `tyk.conf`:

- `use_distributed_counter` to `true`: This enables the feature 
- `distributed_quota_set_master` to `true`: Since we only have redis as a pub/sub mechanism at the moment, we need all nodes to be masters, this causes all nodes to persist the CRDT list and is therefore fault tolerant. This is optional, obviously, if you can be completely sure that your master will not fail - so long as one node is set as master we are clear.

It will work in single-host mode just like a regular quota.

There is a unit test, but multi-host testing was done manually, as well a failure conditions:

*Manual testing procedure:*

This is best done with a docker env because you need to generate a network of gateways.

1. Start 3 gateways against a dashboard and configure an API that uses token auth
2. Set up a gateway in CE mode and create an open API that round robins to the 3 other gateways, we are basically forcing distributed access via the CE gateway
3. Create a token with a quota inthe distributed nodes (dashboard)
4. Send traffic to the CE gateway APi that is round-robinning to the 3 nodes
5. Watch the returned headers of the response, there will be an X-Ratelimit-Remaining header that should count down
6. Ensure that an error is returned once you reach the quota

Note: The countdown is only partially accurate due to the distributed persistence, *however* the quota *will* be enforced accurately.

Further testing should then be the above procedure but with the following changes:

1. Start the cluster, then stop all instances mid-test, then start again
2. Start the cluster, kill one of the 3 nodes and then start it again while sending traffic (the requests will fail when the request hits a dead server, but that doesn;t matter, the quota should count down and continue to be enforced when it remains down AND when it is brought back online).
